### DISCUSSION TITLE
Getting message: "new row violates row-level security policy for table \"objects\"" statusCode: "42501"

### DISCUSSION THREAD

**Discussion by za01br** [supabase/supabase#3389] [Questions] [answered]
Hey all.

What I did:

1. Created an account
2. Created a bucket (public)
3. Created a folder inside the bucket
4. On my code, I try to upload an image:
```
const { data, error } = await supabase.storage
      .from("gpic-all-ima1ges-v1")
      .upload("mvp/file1.png", file, {
        cacheControl: "3600",
      });
```

But I am getting:
`"new row violates row-level security policy for table \"objects\"" statusCode: "42501"`

- I don't have any tables
- I didn't set any policy, because my bucket is set to "public"

Not sure what I am missing here. 😅
I saw one or two questions about this, but they are related to tables/policies.

**Comment by za01br** [comment, marked as answer]
Well, I'm probably the only one that took forever to understand this, but:

- Even if the bucket is "public", you still need to set up policies
- In a case where people should be able to VIEW and UPLOAD files to a bucket, you need to:

1. Go to SQL
2. `+ New Query`
3. Use this query and hit "Run

```
create policy "Public Access Bucket Images Get"
on storage.objects for select
using ( bucket_id = 'YOUR-BUCKET-NAME' );

create policy "Public Access Bucket Images Post"
on storage.objects for insert
with check ( bucket_id = 'YOUR-BUCKET-NAME' );
```

All good now! 🤙🏻

**Reply by Aminigbo** [reply]
I tried this  but its returning an error message of

Error adding policy: syntax error at or near "CREATE"

**Reply by jstdk** [reply]
You saved my weekend. I love Supabase, but the documentation of the Storage sections could be better. I will contribute. I literally have been stuck on saving and reading an avatar table for almost a week.

**Comment by aman246149** [comment]
-- For INSERT operations (uploading files)
CREATE POLICY "Enable insert for authenticated users" ON storage.objects FOR INSERT 
WITH CHECK (bucket_id = YOURBUCKETNAME' AND auth.role() = 'authenticated');

-- For SELECT operations (reading/downloading files)
CREATE POLICY "Enable read access for all users" ON storage.objects FOR SELECT 
USING (bucket_id = 'YOURBUCKETNAME');

-- For UPDATE operations
CREATE POLICY "Enable update for authenticated users" ON storage.objects FOR UPDATE
USING (bucket_id = 'YOURBUCKETNAME' AND auth.role() = 'authenticated')
WITH CHECK (bucket_id = 'YOURBUCKETNAME' AND auth.role() = 'authenticated');

-- For DELETE operations
CREATE POLICY "Enable delete for authenticated users" ON storage.objects FOR DELETE 
USING (bucket_id = 'YOURBUCKETNAME' AND auth.role() = 'authenticated');

**Reply by aman246149** [reply]
write this command on sql editor and run ,it will work then
